libs/net/dal/Services/FileReferenceService.cs

```diff
@@ -254,5 +258,155 @@ public async Task<IEnumerable<FileReference>> GetFiles(DateTime? updatedBefore =
+                BucketName = bucketName,
+                Key = s3Key
+            };
+
+            var response = await s3Client.GetObjectAsync(request);
+
+            if (response.HttpStatusCode == System.Net.HttpStatusCode.OK)
+            {
+                var memoryStream = new MemoryStream();
+                await response.ResponseStream.CopyToAsync(memoryStream);
+                memoryStream.Position = 0;
+                return memoryStream;
+            }
+            else
+            {
+                Logger.LogError("Failed to download file from S3: {S3Key}", s3Key);
+                throw new Exception($"Failed to download file from S3: {s3Key}");
+            }
+        }
+        catch (AmazonS3Exception ex)
+        {
+            Logger.LogError(ex, "Error retrieving file from S3: {S3Key}", s3Key);
+            throw;
+        }
+        catch (Exception ex)
+        {
+            Logger.LogError(ex, "Unexpected error when downloading file from S3: {S3Key}", s3Key);
+            throw;
+        }
+    }
+
+    public async Task<FileReference?> GetByS3PathAsync(string s3Path)
+    {
+        return await this.Context.FileReferences
+            .FirstOrDefaultAsync(fr => fr.S3Path == s3Path);
+    }
+
+    public async Task<FileReference?> GetByPathAsync(string path)
+    {
+        return await this.Context.FileReferences
+            .FirstOrDefaultAsync(fr => fr.Path == path);
+    }
+
+
+    public async Task<(Stream? Stream, string? FileName, string? ContentType)> GetFileStreamAsync(string path)
+    {
+        var fileReference = await GetByS3PathAsync(path);
+        if (fileReference == null)
+        {
+            fileReference = await GetByPathAsync(path);
+        }
+        if (fileReference == null)
+        {
+            Logger.LogInformation("File reference not found for path: {Path}", path);
+            return (null, null, null);
+        }
+
+        if (fileReference.IsSyncedToS3)
+        {
+            try
+            {
+                var stream = await DownloadFromS3Async(path);
+                return (stream, fileReference.FileName, fileReference.ContentType);
+            }
+            catch (Exception ex)
+            {
+                Logger.LogError(ex, "Error on stream file from S3: {Path}", path);
+            }
+        }
+
+        var safePath = Path.Combine(_options.GetUploadPath(), path);
+        if (!File.Exists(safePath))
+        {
+            Logger.LogInformation("File does not exist: {Path}", safePath);
+            return (null, null, null);
+        }
+
+        var fileStream = File.OpenRead(safePath);
+        return (fileStream, fileReference.FileName, fileReference.ContentType);
+    }
+
+    public async Task<FileReference> UpdateAsync(FileReference entity)
+    {
+        this.Context.Update(entity);
+        await this.Context.SaveChangesAsync();
+        return entity;
+    }
+
+    public async Task<int> DeleteOldLocalFilesAsync(DateTime beforeDate)
```

Comment:
I don't think this function makes business sense.  We would never want to delete old local files as a separate process.  We want to migrate files to S3 and at the same time and only after successfully copying the file we would delete the old reference.  This requires an update to the database and indexing to occur too.

Reply:
I still don't know why we would need this specific function.  We only ever want to delete a file when we migrate it from Openshift to S3.